sys/linux/devio.txt

```diff
@@ -0,0 +1,129 @@
+include <linux/usbdevice_fs.h>
+include <uapi/linux/fcntl.h>
+
+define MAXDRIVERNAME	USBDEVFS_MAXDRIVERNAME + 1
+
+resource fd_devio[fd]
+
+syz_open_dev$usb(dev ptr[in, string["/dev/bus/usb/00#/00#"]], id intptr, flags flags[open_flags]) fd_devio
+
+read$devio(fd fd_devio, buf buffer[out], count len[buf])
+
+mmap$devio(addr vma, len len[addr], prot flags[mmap_prot], flags flags[mmap_flags], fd fd_devio, offset fileoff)
+_ = __NR_mmap2
+
+ioctl$USBDEVFS_REAPURB(fd fd_devio, cmd const[USBDEVFS_REAPURB], arg ptr[out, int64])
+ioctl$USBDEVFS_REAPURBNDELAY(fd fd_devio, cmd const[USBDEVFS_REAPURBNDELAY], arg ptr[out, int64])
+ioctl$USBDEVFS_CONTROL(fd fd_devio, cmd const[USBDEVFS_CONTROL], arg ptr[in, usbdevfs_ctrltranfer])
+ioctl$USBDEVFS_BULK(fd fd_devio, cmd const[USBDEVFS_BULK], arg ptr[in, usbdevfs_bulktransfer])
+ioctl$USBDEVFS_RESETEP(fd fd_devio, cmd const[USBDEVFS_RESETEP], arg ptr[in, int32])
+ioctl$USBDEVFS_RESET(fd fd_devio, cmd const[USBDEVFS_RESET])
+ioctl$USBDEVFS_CLEAR_HALT(fd fd_devio, cmd const[USBDEVFS_CLEAR_HALT], arg ptr[in, int32])
+ioctl$USBDEVFS_GETDRIVER(fd fd_devio, cmd const[USBDEVFS_GETDRIVER], arg ptr[inout, usbdevfs_getdriver])
+ioctl$USBDEVFS_CONNECTINFO(fd fd_devio, cmd const[USBDEVFS_CONNECTINFO], arg ptr[out, usbdevfs_connectinfo])
+ioctl$USBDEVFS_SETINTERFACE(fd fd_devio, cmd const[USBDEVFS_SETINTERFACE], arg ptr[in, usbdevfs_setinterface])
+ioctl$USBDEVFS_SETCONFIGURATION(fd fd_devio, cmd const[USBDEVFS_SETCONFIGURATION], arg ptr[in, int32])
+ioctl$USBDEVFS_SUBMITURB(fd fd_devio, cmd const[USBDEVFS_SUBMITURB], arg ptr[in, usbdevfs_urb])
+ioctl$USBDEVFS_DISCARDURB(fd fd_devio, cmd const[USBDEVFS_DISCARDURB], arg ptr[in, int64])
+ioctl$USBDEVFS_DISCSIGNAL(fd fd_devio, cmd const[USBDEVFS_DISCSIGNAL], arg ptr[in, usbdevfs_disconnectsignal])
+ioctl$USBDEVFS_CLAIMINTERFACE(fd fd_devio, cmd const[USBDEVFS_CLAIMINTERFACE], arg ptr[in, int32])
+ioctl$USBDEVFS_RELEASEINTERFACE(fd fd_devio, cmd const[USBDEVFS_RELEASEINTERFACE], arg ptr[in, int32])
+ioctl$USBDEVFS_IOCTL(fd fd_devio, cmd const[USBDEVFS_IOCTL], arg ptr[in, usbdevfs_ioctl])
+ioctl$USBDEVFS_CLAIM_PORT(fd fd_devio, cmd const[USBDEVFS_CLAIM_PORT], arg ptr[in, int32])
+ioctl$USBDEVFS_RELEASE_PORT(fd fd_devio, cmd const[USBDEVFS_RELEASE_PORT], arg ptr[in, int32])
+ioctl$USBDEVFS_GET_CAPABILITIES(fd fd_devio, cmd const[USBDEVFS_GET_CAPABILITIES], arg ptr[out, int32])
+ioctl$USBDEVFS_DISCONNECT_CLAIM(fd fd_devio, cmd const[USBDEVFS_DISCONNECT_CLAIM], arg ptr[in, usbdevfs_disconnect_claim])
+ioctl$USBDEVFS_FREE_STREAMS(fd fd_devio, cmd const[USBDEVFS_FREE_STREAMS], arg ptr[in, usbdevfs_streams])
+ioctl$USBDEVFS_DROP_PRIVILEGES(fd fd_devio, cmd const[USBDEVFS_DROP_PRIVILEGES], arg ptr[in, int32])
+ioctl$USBDEVFS_GET_SPEED(fd fd_devio, cmd const[USBDEVFS_GET_SPEED])
+
+usbdevfs_ctrltranfer {
+	bRequestType	int8
+	bRequest	int8
+	wValue		int16
+	wIndex		int16
+	wLength		int16
+	timeout		int32
+	data		ptr[inout, array[int8]]
+}
+
+usbdevfs_bulktransfer {
+	ep	int32
```

Comment:
Let's split `ep` into bitfields. The higher bit with values `USB_DIR_*`, and the lower 7 bits with values `int32[0:15]` as a USB device has up to 16 endpoints if each direction.